library/lua/time.lua

```diff
@@ -0,0 +1,87 @@
+local _ENV = mkmodule('time')
+
+local TU_PER_DAY = 1200
+--[[
+if advmode then TU_PER_DAY = 86400 ? or only for cur_year_tick?
+advmod_TU / 72 = ticks
+--]]
+local TU_PER_MONTH = TU_PER_DAY * 28
+local TU_PER_YEAR = TU_PER_MONTH * 12
+
+local MONTHS = {
+ 'Granite',
+ 'Slate',
+ 'Felsite',
+ 'Hematite',
+ 'Malachite',
+ 'Galena',
+ 'Limestone',
+ 'Sandstone',
+ 'Timber',
+ 'Moonstone',
+ 'Opal',
+ 'Obsidian',
+}
+
+Time = defclass(Time)
+function Time:init(args)
+ self.year = args.year or 0
+ self.ticks = args.ticks or 0
+end
+function Time:getDays() -- >>float<< Days as age (including years)
+ return self.year * 336 + (self.ticks / TU_PER_DAY)
+end
+function Time:getDayInYear() -- >>int<< Current day of the year
+ return math.floor (self.ticks / TU_PER_DAY)
+end
+function Time:getDayInMonth()
+ return math.floor ( (self.ticks % TU_PER_MONTH) / TU_PER_DAY ) + 1
+end
+function Time:getMonths() -- >>int<< Months as age (not including years)
+ return math.floor (self.ticks / TU_PER_MONTH)
+end
+function Time:getYears() -- >>int<<
```

Comment:
why does getDays return a float and partial days but getYears does not? I'm not exactly sure which style is better, but I think they should at least be consistent.

Reply:
I can't think of a good reason. I lean toward making everything a float for this stuff unless it absolutely makes no sense to do so. The user can always convert to int.

Reply:
I'm thinking use case for this class will want an integer most of the time. But as we know, ints lose data. So to keep it consistent we either have methods for both floats and ints? Or add a parameter to the methods to indicate which type to return. Or use floats to preserve data and let the user convert.

So assuming we keep things simple and use floats, when the user needs ints, the code will look like:

string.format("%03d-%02d-%02d",agr_time:getYears()//1, agr_time:getMonths()//1, agr_time:getDayInMonth()//1)  Or math.floor instead of //1

%.0f isn't a viable substitute due to rounding up and formatting.

These points are probably all obvious. Just thought I'd mention it regardless. Also my prior experience with Lua is close to 0.

Reply:
another option is to return the result in two integer variables, whole and part. e.g.

```lua
function Datetime:getYears()
    return self.years, self.year_tick
end

Reply:
DF consistently manages times as a pair of integers `(years, year_tick)`. i strongly recommend that we assiduously use this format for representing times and time intervals and eschew entirely the use of floating point for this purpose

Reply:
sgtm